### PR TITLE
[#48] Support for MetadataSet and MetadataSetMember.

### PR DESCRIPTION
**Pull request by samuelmale**
Ticket: https://github.com/mekomsolutions/openmrs-module-initializer/issues/48

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/initializer/api/mds/MetadataSetMemberLineProcessor.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import org.openmrs.module.initializer.api.CsvLine;
+import org.openmrs.module.metadatamapping.MetadataSet;
+import org.openmrs.module.metadatamapping.MetadataSetMember;
+import org.openmrs.module.metadatamapping.api.MetadataMappingService;
+import org.springframework.beans.factory.annotation.Autowired;
+
+@OpenmrsProfile(modules = { "metadatamapping:*" })
+public class MetadataSetMemberLineProcessor extends BaseLineProcessor<MetadataSetMember> {
+	
+	protected static String METADATA_SET_ID = "metadata set";
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/mds/MetadataSetMemberLineProcessor.java]:
```java
final public static String METADATA_SET_UUID = "metadata set uuid";
```

Comment (mks-d) [reply api/src/main/java/org/openmrs/module/initializer/api/mds/MetadataSetMemberLineProcessor.java]: Let's make all such headers `final public static` btw.

File: api/src/main/java/org/openmrs/module/initializer/api/mds/MetadataSetMemberLineProcessor.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+		setMember.setMetadataUuid(line.get(METADATA_UUID, true));
+		String metadataSetRef = line.get(METADATA_SET_ID, true);
+		if (StringUtils.isNotBlank(metadataSetRef)) {
+			MetadataSet set = service.getMetadataSetByUuid(metadataSetRef);
+			if (set == null) {
+				throw new IllegalArgumentException("Could not find metadataset identified by: " + metadataSetRef);
+			}
+			setMember.setMetadataSet(set);
+		} else {
+			throw new IllegalArgumentException("Metadata set cannot be null or blank");
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/mds/MetadataSetMemberLineProcessor.java]: I wouldn't do this, I would let the service layer API fail on a `null` reference to a set.

File: api/src/main/java/org/openmrs/module/initializer/api/mds/MetadataSetMemberLineProcessor.java (hunk 1 above)
Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/initializer/api/mds/MetadataSetMemberLineProcessor.java]: Resolved

File: api/src/main/java/org/openmrs/module/initializer/api/mds/MetadataSetMemberLineProcessor.java (hunk 2 above)
Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/initializer/api/mds/MetadataSetMemberLineProcessor.java]: Resolved

File: api/src/main/java/org/openmrs/module/initializer/api/mdm/MetadataTermMappingsLineProcessor.java [hunk 3]
```diff
@@ -11,13 +11,13 @@
 @OpenmrsProfile(modules = { "metadatamapping:*" })
 public class MetadataTermMappingsLineProcessor extends BaseLineProcessor<MetadataTermMapping> {
 	
-	protected static String MAPPING_SOURCE = "Mapping source";
+	final public static String MAPPING_SOURCE = "Mapping source";
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/initializer/api/mdm/MetadataTermMappingsLineProcessor.java]: It doesn't affect the line processors, but typically the header names are lowercased. Such as:
```java
final public static String MAPPING_SOURCE = "mapping source";
```
Like you did in `MetadataSetMemberLineProcessor` basically.

File: api/src/test/java/org/openmrs/module/initializer/api/MetadataSetMembersLoaderIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+		{
+			MetadataSetMember member = new MetadataSetMember();
+			member.setName("Old OpenMRS Id");
+			member.setMetadataClass(PATIENT_IDENTIFIER_TYPE_CLASS_NAME);
+			member.setMetadataSet(metadataSet);
+			member.setMetadataUuid("8f6ed8bb-0cbe-4c67-bc45-c5c0320e1324");
+			member.setUuid(MEMBER_TO_RETIRE);
+			service.saveMetadataSetMember(member);
+		}
+	}
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/MetadataSetMembersLoaderIntegrationTest.java]: There are test data sets to enrich now, see here: https://github.com/mekomsolutions/openmrs-module-initializer/tree/abadd7575bd6381d6a6cebea4aaa97578222496f/api/src/test/resources/testdata

File: api/src/test/java/org/openmrs/module/initializer/api/MetadataSetMembersLoaderIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+	public void load_shouldLoadMetadataSetMembersFromCSVLines() {
+		// Setup
+		MetadataSetMember member = null;
+		
+		// Replay
+		loader.load();
+		
+		// Verify created
+		member = service.getMetadataSetMemberByUuid("dbfd899d-e9e1-4059-8992-73737c924f88");
+		Assert.assertEquals("Out Patient Id", member.getName());
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/MetadataSetMembersLoaderIntegrationTest.java]: A detail but:
```java
"Outpatient ID"
```

File: api/src/test/java/org/openmrs/module/initializer/api/MetadataSetsLoaderIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+			service.saveMetadataSet(set);
+		}
+		// Set for retirement
+		{
+			MetadataSet set = new MetadataSet();
+			set.setName("Secondary Identifier Types");
+			set.setUuid(SET_TO_RETIRE_UUID);
+			service.saveMetadataSet(set);
+		}
+	}
```
Comment (mks-d) [inline api/src/test/java/org/openmrs/module/initializer/api/MetadataSetsLoaderIntegrationTest.java]: Same remark, enrich the appropriate test data set and load it here.

File: api/src/test/java/org/openmrs/module/initializer/api/MetadataSetMembersLoaderIntegrationTest.java (hunk 4 above)
Comment (samuelmale) [reply api/src/test/java/org/openmrs/module/initializer/api/MetadataSetMembersLoaderIntegrationTest.java]: @mks-d Resolved

File: api/src/test/java/org/openmrs/module/initializer/api/MetadataSetsLoaderIntegrationTest.java (hunk 6 above)
Comment (samuelmale) [reply api/src/test/java/org/openmrs/module/initializer/api/MetadataSetsLoaderIntegrationTest.java]: @mks-d Resolved

File: api/src/test/java/org/openmrs/module/initializer/api/MetadataSetMembersLoaderIntegrationTest.java (hunk 5 above)
Comment (samuelmale) [reply api/src/test/java/org/openmrs/module/initializer/api/MetadataSetMembersLoaderIntegrationTest.java]: Resolved

File: api/src/main/java/org/openmrs/module/initializer/api/mdm/MetadataTermMappingsLineProcessor.java (hunk 3 above)
Comment (samuelmale) [reply api/src/main/java/org/openmrs/module/initializer/api/mdm/MetadataTermMappingsLineProcessor.java]: Resolved

**Pull request merged**